apps/src/JSInterpreter.js

```diff
@@ -668,6 +672,36 @@ JSInterpreter.prototype.getProperty = function (
 };
 
 /**
+ * Wrapper to Interpreter's hasProperty (extended for custom marshaling)
+ *
+ * Does the named property exist on a data object.
+ * @param {!Object} interpeter Interpreter instance.
+ * @param {!Function} baseHasProperty Original hasProperty() implementation.
+ * @param {!Object} obj Data object.
+ * @param {*} name Name of property.
+ * @return {boolean} True if property exists.
+ */
+JSInterpreter.prototype.hasProperty = function (
+    interpreter,
+    baseHasProperty,
+    obj,
+    name) {
+  name = name.toString();
+  var nativeParent;
+  if (obj.isCustomMarshal ||
+      (obj === this.globalScope &&
+          (!!(nativeParent = this.customMarshalGlobalProperties[name])))) {
```

Comment:
It's pretty hard to factor in that way. We only need `nativeParent` in one of two cases. I tried going down that path, but you need to create a function with multiple return values (`boolean` for whether or not you should marshal, `nativeParent` for one of the two cases where you should marshal).